File: src/api/storage/storage.js
/**
 * 持久化存储 API — 手动落盘
 *
 * 适合播放列表、乐库索引等大批量数据。
 * 写入仅修改内存缓存，需显式调用 `storageSave()` 持久化到 storage.json。
 *
 * @example
 * import { storageGet, storageSet, storageSave } from '@/api/storage/storage';
 * await storageSet('playlist', playlistData);
 * await storageSave();
 */

import { transport } from '@/api/transport';

/**
 * 读取键值
 * @param {string} key - 键名
 * @returns {Promise<any>} 存储的值
 */
export async function storageGet(key) {
  return transport.command('storage_get', { key });
}

/**
 * 写入键值（仅修改内存缓存，不落盘）
 * @param {string} key - 键名
 * @param {any} value - 要存储的值
 * @returns {Promise<void>}
 */
export async function storageSet(key, value) {
  return transport.command('storage_set', { key, value });
}

/**
 * 删除键
 * @param {string} key - 键名
 * @returns {Promise<boolean>} 是否成功删除
 */
export async function storageRemove(key) {
  return transport.command('storage_remove', { key });
}

/**
 * 检查键是否存在
 * @param {string} key - 键名
 * @returns {Promise<boolean>}
 */
export async function storageHas(key) {
  return transport.command('storage_has', { key });
}

/**
 * 获取所有键
 * @returns {Promise<string[]>}
 */
export async function storageKeys() {
  return transport.command('storage_keys');
}

/**
 * 清空所有数据（仅修改内存缓存）
 * @returns {Promise<void>}
 */
export async function storageClear() {
  return transport.command('storage_clear');
}

/**
 * 将内存数据持久化到磁盘
 * @returns {Promise<void>}
 */
export async function storageSave() {
  return transport.command('storage_save');
}

File: src-tauri/crates/chordial-server/src/routes/storage.rs
//! Storage 域路由 — 手动落盘持久化存储（含 Blob）。
//!
//! | 方法 | 路径 | 对应命令 |
//! |------|------|---------|
//! | GET | `/storage/:key` | `storage_get` |
//! | PUT | `/storage/:key` | `storage_set` |
//! | DELETE | `/storage/:key` | `storage_remove` |
//! | HEAD | `/storage/:key` | `storage_has` |
//! | GET | `/storage` | `storage_keys` |
//! | DELETE | `/storage` | `storage_clear` |
//! | POST | `/storage/save` | `storage_save` |
//! | PUT | `/storage/blob/:key` | `storage_set_blob` (body: bytes) |
//! | GET | `/storage/blob/:key` | `storage_get_blob` |
//! | DELETE | `/storage/blob/:key` | `storage_remove_blob` |
//! | HEAD | `/storage/blob/:key` | `storage_has_blob` |
//! | GET | `/storage/blob` | `storage_blob_keys` |
//! | DELETE | `/storage/blob` | `storage_clear_blobs` |

use crate::state::AppState;
use axum::body::Bytes;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::routing::{get, post};
use axum::{Json, Router};
use serde_json::Value;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/storage", get(keys).delete(clear))
        .route("/storage/save", post(save))
        .route("/storage/:key", get(get_one).put(set).delete(remove).head(has))
        // Blob 存储
        .route("/storage/blob", get(blob_keys).delete(clear_blobs))
        .route(
            "/storage/blob/:key",
            get(blob_get).put(blob_set).delete(blob_remove).head(blob_has),
        )
}

/// `GET /storage/:key` → `storage_get`
async fn get_one(State(state): State<AppState>, Path(key): Path<String>) -> impl IntoResponse {
    match state.ctx.store.get_raw(&key) {
        Some(v) => Ok(Json(v)),
        None => Err((StatusCode::NOT_FOUND, format!("键 '{}' 不存在", key))),
    }
}

/// `PUT /storage/:key` → `storage_set`
async fn set(
    State(state): State<AppState>,
    Path(key): Path<String>,
    Json(value): Json<Value>,
) -> StatusCode {
    state.ctx.store.set_raw(&key, value);
    StatusCode::NO_CONTENT
}

/// `DELETE /storage/:key` → `storage_remove`
async fn remove(State(state): State<AppState>, Path(key): Path<String>) -> Json<bool> {
    Json(state.ctx.store.remove(&key))
}

/// `HEAD /storage/:key` → `storage_has`
async fn has(State(state): State<AppState>, Path(key): Path<String>) -> StatusCode {
    if state.ctx.store.has(&key) {
        StatusCode::OK
    } else {
        StatusCode::NOT_FOUND
    }
}

/// `GET /storage` → `storage_keys`
async fn keys(State(state): State<AppState>) -> Json<Vec<String>> {
    Json(state.ctx.store.keys())
}

/// `DELETE /storage` → `storage_clear`
async fn clear(State(state): State<AppState>) -> StatusCode {
    state.ctx.store.clear();
    StatusCode::NO_CONTENT
}

/// `POST /storage/save` → `storage_save`
async fn save(State(state): State<AppState>) -> Result<StatusCode, String> {
    state.ctx.store.save()?;
    Ok(StatusCode::NO_CONTENT)
}

// ── Blob ────────────────────────────────────────────

/// `PUT /storage/blob/:key` → `storage_set_blob`（raw bytes body）
async fn blob_set(
    State(state): State<AppState>,
    Path(key): Path<String>,
    body: Bytes,
) -> Result<StatusCode, String> {
    state.ctx.store.set_blob(&key, &body)?;
    Ok(StatusCode::NO_CONTENT)
}

/// `GET /storage/blob/:key` → `storage_get_blob`
async fn blob_get(State(state): State<AppState>, Path(key): Path<String>) -> impl IntoResponse {
    match state.ctx.store.get_blob(&key) {
        Some(data) => Ok::<_, (StatusCode, String)>(data),
        None => Err((StatusCode::NOT_FOUND, format!("Blob 存储项 '{}' 不存在", key))),
    }
}

/// `DELETE /storage/blob/:key` → `storage_remove_blob`
async fn blob_remove(State(state): State<AppState>, Path(key): Path<String>) -> Json<bool> {
    Json(state.ctx.store.remove_blob(&key))
}

/// `HEAD /storage/blob/:key` → `storage_has_blob`
async fn blob_has(State(state): State<AppState>, Path(key): Path<String>) -> StatusCode {
    if state.ctx.store.has_blob(&key) {
        StatusCode::OK
    } else {
        StatusCode::NOT_FOUND
    }
}

/// `GET /storage/blob` → `storage_blob_keys`
async fn blob_keys(State(state): State<AppState>) -> Json<Vec<String>> {
    Json(state.ctx.store.blob_keys())
}

/// `DELETE /storage/blob` → `storage_clear_blobs`
async fn clear_blobs(State(state): State<AppState>) -> StatusCode {
    state.ctx.store.clear_blobs();
    StatusCode::NO_CONTENT
}
